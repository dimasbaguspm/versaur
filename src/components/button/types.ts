import type { ButtonHTMLAttributes } from 'react'

/**
 * ButtonProps defines the props for the Button component
 * @property variant - Visual style variant (primary, secondary, etc)
 * @property size - Size of the button (sm, md, lg)
 * @property disabled - Whether the button is disabled
 * @property type - Button type attribute
 * @property onClick - Click event handler
 */
export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * Visual style variant (semantic, outline, ghost, destructive)
   * Supports semantic variants and their outline/ghost forms
   */
  variant?:
    | 'primary'
    | 'primary-outline'
    | 'primary-ghost'
    | 'secondary'
    | 'secondary-outline'
    | 'secondary-ghost'
    | 'tertiary'
    | 'tertiary-outline'
    | 'tertiary-ghost'
    | 'success'
    | 'success-outline'
    | 'success-ghost'
    | 'info'
    | 'info-outline'
    | 'info-ghost'
    | 'warning'
    | 'warning-outline'
    | 'warning-ghost'
    | 'danger'
    | 'danger-outline'
    | 'danger-ghost'
    | 'outline'
    | 'ghost'
    | 'destructive'
  size?: 'sm' | 'md' | 'lg'
  disabled?: boolean
}
