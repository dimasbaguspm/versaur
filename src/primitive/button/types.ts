import type { ButtonHTMLAttributes } from 'react'

/**
 * ButtonProps defines the props for the Button component
 * @property variant - Visual style variant (primary, ghost, outline, destructive)
 * @property size - Size of the button (sm, md, lg)
 * @property disabled - Whether the button is disabled
 * @property busy - Whether the button is in a loading/busy state
 * @property type - Button type attribute (button, submit, reset)
 */
export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /**
   * Visual style variant
   * - primary: Main action button with coral background (default)
   * - ghost: Subtle, minimal visual weight for secondary actions
   * - outline: Bordered lightweight alternative for secondary actions
   * - destructive: For dangerous or irreversible actions (delete, remove, etc.)
   */
  variant?: 'primary' | 'ghost' | 'outline' | 'destructive'
  /**
   * Size of the button
   * - sm: 28px height, compact for space-constrained interfaces
   * - md: 36px height, standard for most use cases (default)
   * - lg: 40px height, prominent for primary actions
   */
  size?: 'sm' | 'md' | 'lg'
  /**
   * Whether the button is disabled
   * When true, the button becomes non-interactive and visually dimmed
   */
  disabled?: boolean

  /**
   * Whether the button is in a loading/busy state
   * Can be used to show loading indicators or disable interaction during async operations
   */
  busy?: boolean
  /**
   * Button type attribute
   * - button: Standard button (default, prevents form submission)
   * - submit: Submits the containing form
   * - reset: Resets the containing form
   */
  type?: 'button' | 'submit' | 'reset'
}
