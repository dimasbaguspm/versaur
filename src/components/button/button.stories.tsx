import type { Meta, StoryObj } from '@storybook/react'
import { Button } from './button'

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: `
The Button component is a versatile, accessible button with support for semantic variants, outline styles, ghost styles, and multiple sizes. Built with Tailwind CSS and following Versaur's warm, earthy design system.

## Features
- **Semantic Variants**: Primary, secondary, tertiary, success, info, warning, danger
- **Style Variants**: Filled, outline, ghost
- **Sizes**: Small, medium, large
- **Accessibility**: Full keyboard navigation, ARIA attributes, focus management
- **Responsive**: Mobile-first responsive design

## Color Harmony
Uses Versaur's harmonious color palette with warm, earthy tones that create natural visual flow and excellent accessibility.
        `,
      },
    },
  },
  argTypes: {
    variant: {
      control: { type: 'select' },
      options: [
        'primary',
        'primary-outline',
        'primary-ghost',
        'secondary',
        'secondary-outline',
        'secondary-ghost',
        'tertiary',
        'tertiary-outline',
        'tertiary-ghost',
        'success',
        'success-outline',
        'success-ghost',
        'info',
        'info-outline',
        'info-ghost',
        'warning',
        'warning-outline',
        'warning-ghost',
        'danger',
        'danger-outline',
        'danger-ghost',
        'outline',
        'ghost',
        'destructive',
      ],
      description: 'Visual style variant of the button',
    },
    size: {
      control: { type: 'select' },
      options: ['sm', 'md', 'lg'],
      description: 'Size of the button',
    },
    disabled: {
      control: { type: 'boolean' },
      description: 'Whether the button is disabled',
    },
    children: {
      control: { type: 'text' },
      description: 'Button content',
    },
  },
}

export default meta
type Story = StoryObj<typeof Button>

// =============================================================================
// Color Stories - Semantic color system
// =============================================================================

/**
 * Complete color matrix showing all semantic colors in all style variants.
 * Demonstrates the comprehensive color system and variant combinations.
 */
export const ColorMatrix: Story = {
  render: () => (
    <div className='space-y-6'>
      <div>
        <h3 className='text-lg font-semibold mb-3'>Filled Variants</h3>
        <div className='flex flex-wrap gap-3'>
          <Button variant='primary'>Primary</Button>
          <Button variant='secondary'>Secondary</Button>
          <Button variant='tertiary'>Tertiary</Button>
          <Button variant='success'>Success</Button>
          <Button variant='info'>Info</Button>
          <Button variant='warning'>Warning</Button>
          <Button variant='danger'>Danger</Button>
        </div>
      </div>

      <div>
        <h3 className='text-lg font-semibold mb-3'>Outline Variants</h3>
        <div className='flex flex-wrap gap-3'>
          <Button variant='primary-outline'>Primary</Button>
          <Button variant='secondary-outline'>Secondary</Button>
          <Button variant='tertiary-outline'>Tertiary</Button>
          <Button variant='success-outline'>Success</Button>
          <Button variant='info-outline'>Info</Button>
          <Button variant='warning-outline'>Warning</Button>
          <Button variant='danger-outline'>Danger</Button>
        </div>
      </div>

      <div>
        <h3 className='text-lg font-semibold mb-3'>Ghost Variants</h3>
        <div className='flex flex-wrap gap-3'>
          <Button variant='primary-ghost'>Primary</Button>
          <Button variant='secondary-ghost'>Secondary</Button>
          <Button variant='tertiary-ghost'>Tertiary</Button>
          <Button variant='success-ghost'>Success</Button>
          <Button variant='info-ghost'>Info</Button>
          <Button variant='warning-ghost'>Warning</Button>
          <Button variant='danger-ghost'>Danger</Button>
        </div>
      </div>
    </div>
  ),
}

/**
 * All semantic colors in their filled variant.
 * Demonstrates the warm, earthy color palette and harmonious relationships.
 */

export const SemanticColors: Story = {
  render: () => (
    <div className='flex flex-wrap gap-4'>
      <Button variant='primary'>Primary</Button>
      <Button variant='secondary'>Secondary</Button>
      <Button variant='tertiary'>Tertiary</Button>
      <Button variant='success'>Success</Button>
      <Button variant='info'>Info</Button>
      <Button variant='warning'>Warning</Button>
      <Button variant='danger'>Danger</Button>
    </div>
  ),
}

/**
 * Outline variants provide a lighter visual weight while maintaining semantic meaning.
 * Perfect for secondary actions or when you need multiple buttons in close proximity.
 */
export const OutlineVariants: Story = {
  render: () => (
    <div className='flex flex-wrap gap-4'>
      <Button variant='primary-outline'>Primary Outline</Button>
      <Button variant='secondary-outline'>Secondary Outline</Button>
      <Button variant='tertiary-outline'>Tertiary Outline</Button>
      <Button variant='success-outline'>Success Outline</Button>
      <Button variant='info-outline'>Info Outline</Button>
      <Button variant='warning-outline'>Warning Outline</Button>
      <Button variant='danger-outline'>Danger Outline</Button>
    </div>
  ),
}

/**
 * Ghost variants are the most subtle, perfect for navigation or when you need
 * a button that doesn't visually compete with other content.
 */
export const GhostVariants: Story = {
  render: () => (
    <div className='flex flex-wrap gap-4'>
      <Button variant='primary-ghost'>Primary Ghost</Button>
      <Button variant='secondary-ghost'>Secondary Ghost</Button>
      <Button variant='tertiary-ghost'>Tertiary Ghost</Button>
      <Button variant='success-ghost'>Success Ghost</Button>
      <Button variant='info-ghost'>Info Ghost</Button>
      <Button variant='warning-ghost'>Warning Ghost</Button>
      <Button variant='danger-ghost'>Danger Ghost</Button>
    </div>
  ),
}

/**
 * Comparison of all three style variants for the primary semantic color.
 * Shows how the same semantic meaning can be expressed with different visual weights.
 */
export const StyleComparison: Story = {
  render: () => (
    <div className='flex flex-wrap gap-4'>
      <Button variant='primary'>Primary Filled</Button>
      <Button variant='primary-outline'>Primary Outline</Button>
      <Button variant='primary-ghost'>Primary Ghost</Button>
    </div>
  ),
}

// =============================================================================
// Size Stories - Different button sizes
// =============================================================================

/**
 * Size comparison showing all three sizes side by side.
 * Demonstrates the proportional scaling of button sizes.
 */
export const SizeComparison: Story = {
  render: () => (
    <div className='flex flex-wrap items-center gap-4'>
      <Button size='sm'>Small</Button>
      <Button size='md'>Medium</Button>
      <Button size='lg'>Large</Button>
    </div>
  ),
}

// =============================================================================
// State Stories - Interactive states and conditions
// =============================================================================

/**
 * Comparison of enabled vs disabled states across different variants.
 * Shows how disabled state affects visual appearance while maintaining accessibility.
 */
export const DisabledComparison: Story = {
  render: () => (
    <div className='space-y-4'>
      <div className='flex flex-wrap gap-4'>
        <Button variant='primary'>Enabled</Button>
        <Button variant='primary' disabled>
          Disabled
        </Button>
      </div>
      <div className='flex flex-wrap gap-4'>
        <Button variant='primary-outline'>Enabled Outline</Button>
        <Button variant='primary-outline' disabled>
          Disabled Outline
        </Button>
      </div>
      <div className='flex flex-wrap gap-4'>
        <Button variant='primary-ghost'>Enabled Ghost</Button>
        <Button variant='primary-ghost' disabled>
          Disabled Ghost
        </Button>
      </div>
    </div>
  ),
}

/**
 * Loading state example showing how buttons can indicate ongoing processes.
 * While not built into the component, shows common usage patterns.
 */
export const LoadingState: Story = {
  args: {
    children: 'Loading...',
    disabled: true,
  },
}
