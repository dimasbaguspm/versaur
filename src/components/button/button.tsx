import React from 'react'
import { cn } from '@/utils/cn'
import { cva } from '@/utils/variants'
import type { ButtonProps } from './types'

const buttonVariants = cva(
  'inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none',
  {
    variants: {
      variant: {
        // Core variants using Versaur color system
        primary:
          'bg-coral text-white hover:bg-coral/90 focus-visible:ring-coral-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        secondary:
          'bg-sage text-white hover:bg-sage/90 focus-visible:ring-sage-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        tertiary:
          'bg-mist text-white hover:bg-mist/90 focus-visible:ring-mist-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        ghost:
          'bg-white text-foreground hover:bg-slate-soft focus-visible:ring-slate-soft focus-visible:ring-offset-white',
        neutral:
          'bg-neutral text-foreground border border-border hover:bg-neutral/80 focus-visible:ring-foreground-soft focus-visible:ring-offset-white shadow-sm',

        // Outline variants
        'primary-outline':
          'border border-coral text-coral bg-white hover:bg-coral hover:text-white focus-visible:ring-coral-soft focus-visible:ring-offset-white transition-all',
        'secondary-outline':
          'border border-sage text-sage bg-white hover:bg-sage hover:text-white focus-visible:ring-sage-soft focus-visible:ring-offset-white transition-all',
        'tertiary-outline':
          'border border-mist text-mist bg-white hover:bg-mist hover:text-white focus-visible:ring-mist-soft focus-visible:ring-offset-white transition-all',
        'ghost-outline':
          'border border-slate text-slate bg-white hover:bg-slate hover:text-white focus-visible:ring-slate-soft focus-visible:ring-offset-white transition-all',
        'neutral-outline':
          'border border-neutral text-foreground bg-white hover:bg-neutral hover:text-foreground focus-visible:ring-foreground-soft focus-visible:ring-offset-white transition-all',

        // Ghost variants (subtle)
        'primary-ghost':
          'text-coral bg-white hover:bg-coral-soft focus-visible:ring-coral-soft focus-visible:ring-offset-white',
        'secondary-ghost':
          'text-sage bg-white hover:bg-sage-soft focus-visible:ring-sage-soft focus-visible:ring-offset-white',
        'tertiary-ghost':
          'text-mist bg-white hover:bg-mist-soft focus-visible:ring-mist-soft focus-visible:ring-offset-white',
        'neutral-ghost':
          'text-foreground bg-white hover:bg-neutral/70 focus-visible:ring-foreground-soft focus-visible:ring-offset-white',

        // Semantic variants
        success:
          'bg-success text-white hover:bg-success/90 focus-visible:ring-success-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'success-outline':
          'border border-success text-success bg-white hover:bg-success hover:text-white focus-visible:ring-success-soft focus-visible:ring-offset-white transition-all',
        'success-ghost':
          'text-success bg-white hover:bg-success-soft focus-visible:ring-success-soft focus-visible:ring-offset-white',

        info: 'bg-info text-white hover:bg-info/90 focus-visible:ring-info-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'info-outline':
          'border border-info text-info bg-white hover:bg-info hover:text-white focus-visible:ring-info-soft focus-visible:ring-offset-white transition-all',
        'info-ghost':
          'text-info bg-white hover:bg-info-soft focus-visible:ring-info-soft focus-visible:ring-offset-white',

        warning:
          'bg-warning text-white hover:bg-warning/90 focus-visible:ring-warning-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'warning-outline':
          'border border-warning text-warning bg-white hover:bg-warning hover:text-white focus-visible:ring-warning-soft focus-visible:ring-offset-white transition-all',
        'warning-ghost':
          'text-warning bg-white hover:bg-warning-soft focus-visible:ring-warning-soft focus-visible:ring-offset-white',

        danger:
          'bg-danger text-white hover:bg-danger/90 focus-visible:ring-danger-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'danger-outline':
          'border border-danger text-danger bg-white hover:bg-danger hover:text-white focus-visible:ring-danger-soft focus-visible:ring-offset-white transition-all',
        'danger-ghost':
          'text-danger bg-white hover:bg-danger-soft focus-visible:ring-danger-soft focus-visible:ring-offset-white',

        // Utility variants
        outline:
          'border border-border text-foreground bg-white hover:bg-accent-soft focus-visible:ring-accent-soft focus-visible:ring-offset-white transition-all',
        destructive:
          'bg-danger text-white hover:bg-danger/90 focus-visible:ring-danger-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
      },
      size: {
        sm: 'h-8 px-3 text-sm min-w-[2.25rem]',
        md: 'h-10 px-4 text-sm min-w-[2.5rem]',
        lg: 'h-12 px-8 text-lg min-w-[2.75rem]',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'md',
    },
  }
)

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      className,
      variant = 'primary',
      size = 'md',
      disabled = false,
      type = 'button',
      children,
      ...props
    },
    ref
  ) {
    return (
      <button
        ref={ref}
        type={type}
        className={cn(buttonVariants({ variant, size }), className)}
        disabled={disabled}
        aria-disabled={disabled}
        inert={disabled ? true : undefined}
        {...props}
      >
        {children}
      </button>
    )
  }
)
