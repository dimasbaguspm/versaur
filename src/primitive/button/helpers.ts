import { cva } from 'class-variance-authority'

export const buttonVariants = cva(
  'inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer',
  {
    variants: {
      variant: {
        // Core variants using Versaur color system
        primary:
          'bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        secondary:
          'bg-secondary text-white hover:bg-secondary/90 focus-visible:ring-secondary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        tertiary:
          'bg-tertiary text-white hover:bg-tertiary/90 focus-visible:ring-tertiary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        ghost:
          'bg-white text-foreground hover:bg-ghost-soft focus-visible:ring-ghost-light focus-visible:ring-offset-white',
        neutral:
          'bg-neutral text-foreground border border-border hover:bg-neutral/80 focus-visible:ring-foreground-light focus-visible:ring-offset-white shadow-sm',

        // Outline variants
        'primary-outline':
          'border border-primary text-primary bg-white hover:bg-primary hover:text-white focus-visible:ring-primary-light focus-visible:ring-offset-white transition-all',
        'secondary-outline':
          'border border-secondary text-secondary bg-white hover:bg-secondary hover:text-white focus-visible:ring-secondary-light focus-visible:ring-offset-white transition-all',
        'tertiary-outline':
          'border border-tertiary text-tertiary bg-white hover:bg-tertiary hover:text-white focus-visible:ring-tertiary-light focus-visible:ring-offset-white transition-all',
        'ghost-outline':
          'border border-ghost text-ghost bg-white hover:bg-ghost hover:text-white focus-visible:ring-ghost-light focus-visible:ring-offset-white transition-all',
        'neutral-outline':
          'border border-border text-foreground bg-white hover:bg-neutral hover:text-foreground focus-visible:ring-foreground-light focus-visible:ring-offset-white transition-all',

        // Ghost variants (subtle)
        'primary-ghost':
          'text-primary bg-white hover:bg-primary/20 focus-visible:ring-primary focus-visible:ring-offset-white',
        'secondary-ghost':
          'text-secondary bg-white hover:bg-secondary/20 focus-visible:ring-secondary focus-visible:ring-offset-white',
        'tertiary-ghost':
          'text-tertiary bg-white hover:bg-tertiary/20 focus-visible:ring-tertiary focus-visible:ring-offset-white',
        'neutral-ghost':
          'text-foreground bg-white hover:bg-neutral/50 focus-visible:ring-foreground focus-visible:ring-offset-white',

        // Semantic variants
        success:
          'bg-success text-white hover:bg-success/90 focus-visible:ring-success-light focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'success-outline':
          'border border-success text-success bg-white hover:bg-success hover:text-white focus-visible:ring-success-light focus-visible:ring-offset-white transition-all',
        'success-ghost':
          'text-success bg-white hover:bg-success/20 focus-visible:ring-success-light focus-visible:ring-offset-white',

        info: 'bg-info text-white hover:bg-info/90 focus-visible:ring-info-light focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'info-outline':
          'border border-info text-info bg-white hover:bg-info hover:text-white focus-visible:ring-info-light focus-visible:ring-offset-white transition-all',
        'info-ghost':
          'text-info bg-white hover:bg-info/20 focus-visible:ring-info-light focus-visible:ring-offset-white',

        warning:
          'bg-warning text-white hover:bg-warning/90 focus-visible:ring-warning-light focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'warning-outline':
          'border border-warning text-warning bg-white hover:bg-warning hover:text-white focus-visible:ring-warning-light focus-visible:ring-offset-white transition-all',
        'warning-ghost':
          'text-warning bg-white hover:bg-warning/20 focus-visible:ring-warning-light focus-visible:ring-offset-white',

        danger:
          'bg-danger text-white hover:bg-danger/90 focus-visible:ring-danger-light focus-visible:ring-offset-white shadow-sm hover:shadow-md',
        'danger-outline':
          'border border-danger text-danger bg-white hover:bg-danger hover:text-white focus-visible:ring-danger-light focus-visible:ring-offset-white transition-all',
        'danger-ghost':
          'text-danger bg-white hover:bg-danger/20 focus-visible:ring-danger-light focus-visible:ring-offset-white',

        // Utility variants
        outline:
          'border border-border text-foreground bg-white hover:bg-accent-soft focus-visible:ring-accent-soft focus-visible:ring-offset-white transition-all',
        destructive:
          'bg-danger text-white hover:bg-danger/90 focus-visible:ring-danger-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md',
      },
      size: {
        sm: 'h-7 px-3 text-sm min-w-[2.25rem]',
        md: 'h-9 px-4 text-sm min-w-[2.5rem]',
        lg: 'h-10 px-8 text-lg min-w-[2.75rem]',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'md',
    },
  }
)
