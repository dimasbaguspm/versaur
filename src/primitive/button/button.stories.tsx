import type { Meta, StoryObj } from '@storybook/react'
import { Button } from './button'

const meta: Meta<typeof Button> = {
  title: 'Primitive/Button',
  component: Button,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: `
The Button component is a simple, accessible button that directly aligns with the native HTML button element. It provides 4 essential variants for different use cases and 3 size options.

## Features
- **4 Variants**: Primary (main actions), ghost (subtle), outline (secondary), destructive (dangerous actions)
- **3 Sizes**: Small, medium (default), large
- **Accessibility**: Full keyboard navigation, ARIA attributes, focus management
- **Browser-aligned**: Behaves like native button with enhanced styling

## Usage
Use the Button component for any clickable action in your UI. Choose the variant based on the action's importance and context:
- **Primary**: Main call-to-action buttons
- **Ghost**: Subtle actions that shouldn't compete visually
- **Outline**: Secondary actions or alternative choices
- **Destructive**: Delete, remove, or other irreversible actions
        `,
      },
    },
  },
  argTypes: {
    variant: {
      control: 'select',
      options: ['primary', 'ghost', 'outline', 'destructive'],
    },
    size: {
      control: 'select',
      options: ['sm', 'md', 'lg'],
    },
    disabled: {
      control: 'boolean',
    },
    busy: {
      control: 'boolean',
    },
  },
}

export default meta
type Story = StoryObj<typeof Button>

// =============================================================================
// Interactive Playground
// =============================================================================

/**
 * Interactive playground to test all button properties.
 * Use controls to change variant, size, disabled, and busy states.
 */
export const Playground: Story = {
  args: {
    children: 'Click me',
    variant: 'primary',
    size: 'md',
    disabled: false,
    busy: false,
  },
}

// =============================================================================
// Variants
// =============================================================================

/**
 * All 4 button variants showing visual hierarchy and use cases.
 * Primary for main actions, ghost for subtle actions, outline for secondary, destructive for dangerous actions.
 */
export const Variants: Story = {
  render: () => (
    <div className='flex flex-wrap gap-4'>
      <Button variant='primary'>Primary</Button>
      <Button variant='ghost'>Ghost</Button>
      <Button variant='outline'>Outline</Button>
      <Button variant='destructive'>Destructive</Button>
    </div>
  ),
}

// =============================================================================
// Sizes
// =============================================================================

/**
 * All three sizes showing proportional scaling.
 * Small (28px), medium/default (36px), and large (40px) heights.
 */
export const Sizes: Story = {
  render: () => (
    <div className='flex flex-wrap items-center gap-4'>
      <Button size='sm'>Small</Button>
      <Button size='md'>Medium</Button>
      <Button size='lg'>Large</Button>
    </div>
  ),
}

// =============================================================================
// States
// =============================================================================

/**
 * All button states: normal, busy (loading), and disabled.
 * Busy state shows spinner and auto-disables interaction. Disabled state reduces opacity.
 */
export const States: Story = {
  render: () => (
    <div className='space-y-4'>
      <div className='flex flex-wrap gap-4'>
        <Button variant='primary'>Normal</Button>
        <Button variant='primary' busy>
          Busy
        </Button>
        <Button variant='primary' disabled>
          Disabled
        </Button>
      </div>
      <div className='flex flex-wrap gap-4'>
        <Button variant='outline'>Normal</Button>
        <Button variant='outline' busy>
          Busy
        </Button>
        <Button variant='outline' disabled>
          Disabled
        </Button>
      </div>
      <div className='flex flex-wrap gap-4'>
        <Button variant='destructive'>Normal</Button>
        <Button variant='destructive' busy>
          Busy
        </Button>
        <Button variant='destructive' disabled>
          Disabled
        </Button>
      </div>
    </div>
  ),
}

// =============================================================================
// Real-world Examples
// =============================================================================

/**
 * Common form actions pattern with cancel and submit buttons.
 */
export const FormActions: Story = {
  render: () => (
    <div className='flex gap-3'>
      <Button variant='outline'>Cancel</Button>
      <Button variant='primary'>Submit</Button>
    </div>
  ),
}

/**
 * Form submission with loading state - typical async operation pattern.
 */
export const FormSubmitting: Story = {
  render: () => (
    <div className='flex gap-3'>
      <Button variant='outline' disabled>
        Cancel
      </Button>
      <Button variant='primary' busy>
        Submitting...
      </Button>
    </div>
  ),
}

/**
 * Destructive action with confirmation pattern for dangerous operations.
 */
export const DestructiveAction: Story = {
  render: () => (
    <div className='flex gap-3'>
      <Button variant='outline'>Cancel</Button>
      <Button variant='destructive'>Delete Account</Button>
    </div>
  ),
}

/**
 * Multiple button sizes working together in context.
 */
export const SizeVariations: Story = {
  render: () => (
    <div className='flex flex-wrap items-center gap-3'>
      <Button size='sm' variant='outline'>
        Cancel
      </Button>
      <Button size='md' variant='primary'>
        Save
      </Button>
      <Button size='lg' variant='primary'>
        Continue
      </Button>
    </div>
  ),
}
