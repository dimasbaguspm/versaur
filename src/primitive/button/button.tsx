import React from 'react'
import { cn } from '@/utils/cn'
import type { ButtonProps } from './types'
import { buttonVariants } from './helpers'
import { Icon } from '../icon'
import { LoaderCircleIcon } from 'lucide-react'

/**
 * Button component following browser-standard button behavior
 * Supports 4 variants: primary, ghost, outline, and destructive
 * Provides 3 sizes: sm, md (default), lg
 */
export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  function Button(
    {
      className,
      variant = 'primary',
      size = 'md',
      disabled = false,
      type = 'button',
      busy,
      children,
      ...props
    },
    ref
  ) {
    // Map button sizes to icon sizes
    const iconSize = size === 'sm' ? 'sm' : size === 'lg' ? 'md' : 'sm'

    // Determine icon color based on variant
    const getIconColor = () => {
      if (variant === 'primary' || variant === 'destructive') {
        return 'white'
      }
      if (variant === 'outline' || variant === 'ghost') {
        return 'inherit'
      }
      return 'inherit'
    }

    return (
      <button
        ref={ref}
        type={type}
        className={cn(buttonVariants({ variant, size }), className)}
        disabled={disabled || busy}
        aria-disabled={disabled || busy}
        inert={disabled || busy ? true : undefined}
        aria-busy={busy ? true : undefined}
        {...props}
      >
        {busy && (
          <Icon
            as={LoaderCircleIcon}
            size={iconSize}
            color={getIconColor()}
            className='animate-spin'
          />
        )}
        {children}
      </button>
    )
  }
)
