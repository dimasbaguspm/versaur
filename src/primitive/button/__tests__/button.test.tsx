import { describe, it, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { Button } from '../button'

import * as stories from '../button.stories'
import { composeStories } from '@storybook/react'

const { Variants, Sizes, States } = composeStories(stories)

describe('Button', () => {
  it('renders correctly', () => {
    render(<Button>Click me</Button>)
    expect(screen.getByRole('button')).toBeInTheDocument()
    expect(screen.getByText('Click me')).toBeInTheDocument()
  })

  it('applies variant classes correctly', () => {
    const { rerender } = render(<Button variant='primary'>Primary</Button>)
    let button = screen.getByRole('button')
    expect(button).toHaveClass('bg-primary')

    rerender(<Button variant='ghost'>Ghost</Button>)
    button = screen.getByRole('button')
    expect(button).toHaveClass('bg-white')

    rerender(<Button variant='outline'>Outline</Button>)
    button = screen.getByRole('button')
    expect(button).toHaveClass('border')

    rerender(<Button variant='destructive'>Destructive</Button>)
    button = screen.getByRole('button')
    expect(button).toHaveClass('bg-danger')
  })

  it('applies size classes correctly', () => {
    const { rerender } = render(<Button size='sm'>Small</Button>)
    let button = screen.getByRole('button')
    expect(button).toHaveClass('h-7')

    rerender(<Button size='md'>Medium</Button>)
    button = screen.getByRole('button')
    expect(button).toHaveClass('h-9')

    rerender(<Button size='lg'>Large</Button>)
    button = screen.getByRole('button')
    expect(button).toHaveClass('h-10')
  })

  it('handles disabled state', () => {
    render(<Button disabled>Disabled</Button>)
    const button = screen.getByRole('button')
    expect(button).toBeDisabled()
    expect(button).toHaveClass('disabled:opacity-50')
    expect(button).toHaveAttribute('aria-disabled', 'true')
  })

  it('handles busy state', () => {
    render(<Button busy>Loading</Button>)
    const button = screen.getByRole('button')
    expect(button).toBeDisabled() // Button should be disabled when busy
    expect(button).toHaveAttribute('aria-busy', 'true')
    // Check for loader icon presence (SVG element)
    const svg = button.querySelector('svg')
    expect(svg).toBeInTheDocument()
    expect(svg).toHaveClass('animate-spin')
  })

  it('accepts custom className', () => {
    render(<Button className='custom-class'>Custom</Button>)
    const button = screen.getByRole('button')
    expect(button).toHaveClass('custom-class')
  })

  it('renders story: Variants', () => {
    render(<Variants />)
    const buttons = screen.getAllByRole('button')
    expect(buttons).toHaveLength(4) // 4 variants: primary, ghost, outline, destructive
    expect(buttons[0]).toHaveTextContent('Primary')
    expect(buttons[1]).toHaveTextContent('Ghost')
    expect(buttons[2]).toHaveTextContent('Outline')
    expect(buttons[3]).toHaveTextContent('Destructive')
  })

  it('renders story: Sizes', () => {
    render(<Sizes />)
    const buttons = screen.getAllByRole('button')
    expect(buttons).toHaveLength(3) // 3 sizes: sm, md, lg
    expect(buttons[0]).toHaveClass('h-7') // sm
    expect(buttons[1]).toHaveClass('h-9') // md
    expect(buttons[2]).toHaveClass('h-10') // lg
  })

  it('renders story: States', () => {
    render(<States />)
    const buttons = screen.getAllByRole('button')
    expect(buttons).toHaveLength(9) // 3 variants Ã— 3 states (normal, busy, disabled)

    // Check for busy states with spinner
    const busyButtons = buttons.filter(btn => btn.hasAttribute('aria-busy'))
    expect(busyButtons).toHaveLength(3)

    // Check for disabled states
    const disabledButtons = buttons.filter(
      btn => btn.hasAttribute('disabled') && !btn.hasAttribute('aria-busy')
    )
    expect(disabledButtons).toHaveLength(3)
  })

  it('matches snapshot', () => {
    const { asFragment } = render(<Button>Snapshot Test</Button>)
    expect(asFragment()).toMatchSnapshot()
  })
})
