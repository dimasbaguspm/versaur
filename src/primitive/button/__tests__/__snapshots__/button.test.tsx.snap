// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md h-9 px-4 text-sm min-w-[2.5rem]"
    type="button"
  >
    Snapshot Test
  </button>
</DocumentFragment>
`;
